src/source/blockedsource.js

```diff
@@ -0,0 +1,239 @@
+import LRUCache from 'lru-cache';
+import { BaseSource } from './basesource';
+import { wait } from '../utils';
+
+class Block {
+  /**
+   *
+   * @param {number} offset
+   * @param {number} length
+   * @param {ArrayBuffer} [data]
+   */
+  constructor(offset, length, data = null) {
+    this.offset = offset;
+    this.length = length;
+    this.data = data;
+  }
+
+  /**
+   * @returns {number} the top byte border
+   */
+  get top() {
+    return this.offset + this.length;
+  }
+}
+
+
+class BlockGroup {
+  /**
+   *
+   * @param {number} offset
+   * @param {number} length
+   * @param {number[]} blockIds
+   */
+  constructor(offset, length, blockIds) {
+    this.offset = offset;
+    this.length = length;
+    this.blockIds = blockIds;
+  }
+}
+
+
+export class BlockedSource extends BaseSource {
+  /**
+   *
+   * @param {Source} source The underlying source that shall be blocked and cached
+   * @param {object} options
+   */
+  constructor(source, { blockSize = 65536, cacheSize = 100 } = {}) {
+    super();
+    this.source = source;
+    this.blockSize = blockSize;
+
+    this.blockCache = new LRUCache({ max: cacheSize });
+
+    // mapping blockId -> Block instance
+    this.blockRequests = new Map();
+
+    // set of blockIds missing for the current requests
+    this.blockIdsToFetch = new Set();
+  }
+
+  get fileSize() {
+    return this.source.fileSize;
+  }
+
+  /**
+   *
+   * @param {./basesource/Slice[]} slices
+   */
+  async fetch(slices, signal) {
+    const cachedBlocks = new Map();
+    const blockRequests = new Map();
+    const missingBlockIds = new Set();
+
+    for (const { offset, length } of slices) {
+      let top = offset + length;
+
+      const { fileSize } = this;
+      if (fileSize !== null) {
+        top = Math.min(top, fileSize);
+      }
+
+      const firstBlockOffset = Math.floor(offset / this.blockSize) * this.blockSize;
+
+      // chunk the current slice into blocks
+      for (let current = firstBlockOffset; current < top; current += this.blockSize) {
+        // check if the block is cached, being requested or still missing
+        const blockId = Math.floor(current / this.blockSize);
+
+        if (this.blockCache.has(blockId)) {
+          cachedBlocks.set(blockId, this.blockCache.get(blockId));
+        } else if (this.blockRequests.has(blockId)) {
+          blockRequests.set(blockId, this.blockRequests.get(blockId));
+        } else {
+          this.blockIdsToFetch.add(blockId);
+          missingBlockIds.add(blockId);
+        }
+      }
+    }
+
+    // allow additional block requests to accumulate
+    await wait();
+
+    // check if we still need to
+    if (this.blockIdsToFetch.size > 0) {
+      const groups = this.groupBlocks(this.blockIdsToFetch);
+
+      // start requesting slices of data
+      const groupRequests = this.source.fetch(groups, signal);
```

Comment:
I think we need handling analogous here to 
https://github.com/geotiffjs/geotiff.js/pull/182/files#diff-908b5c4d8c419c1ab5cd8a06a28e9e47a8e1d1b9b49c8309198ab1e663d54f2fR234

https://github.com/geotiffjs/geotiff.js/pull/182/files#diff-908b5c4d8c419c1ab5cd8a06a28e9e47a8e1d1b9b49c8309198ab1e663d54f2fR264-R268

and 

https://github.com/geotiffjs/geotiff.js/pull/182/files#diff-908b5c4d8c419c1ab5cd8a06a28e9e47a8e1d1b9b49c8309198ab1e663d54f2fR293-R321

for the `AbortSignal` because multiple `fetch` calls can request the same block and then independently abort them.  Without the handling, it seems like an application can lock up like this because they are requesting blocks with one `fetch` that another `fetch` already cancelled but never deleted from the `blockRequests` cache.  Thus the `fetch` that wants the block thinks it is possible to get it but it is in fact not because the request was never satisfied on account of the other `fetch` cancelling it.  I am more than happy to make a PR into this PR to handle that.

![ezgif com-video-to-gif(11)](https://user-images.githubusercontent.com/43999641/108390264-7d12e200-71de-11eb-99a9-5cec0346afc1.gif)

Reply:
Sorry if this is confusing!

Reply:
No, I think I understand the issue, thanks for bringing that up! It basically means that we have some proper error handling at the block level. But I think, this is just a matter of try/catch 🤔 

What is the app you are using? Maybe I can use that to debug the issue.

Reply:
@ilan-gold Could you try again with 07ed101? I think this should solve the issue.

Reply:
@constantinius I'll try it now - I need to get on European-time so we can get through this faster.......need that vaccine....

In any case, I am not sure I have ever shared this with you (I think I have but maybe not?) but we are using geotiff.js extensively in https://github.com/hms-dbmi/viv.  If you run the dev instructions, after `npm start` you'll be using geotiff.js if your browser `image_url` parameter has the `.tiff` ending.  I would recommend http://localhost:8080/?image_url=https://viv-demo.storage.googleapis.com/Vanderbilt-Spraggins-Kidney-MxIF.ome.tif because it is not LZW compressed (we are currently partly on my branch to support that we are using Trevor's LZW library: see https://github.com/geotiffjs/geotiff.js/issues/172).

You'll also need to change what branch you're using in the `package.json` but other than that, you should be able to edit the `node_modules/geotiff.js/src` directly and get live updates.

Reply:
P.S I believe you first need to run `npm uninstall geotiff`

Reply:
@constantinius It is still possible to for independent fetch requests to cancel each others' blocks and not recover.  

Here's an example: If `fetch A` requests `block 1`, it is the only request responsible for fetching it and adds it to `missingBlockIds`. If `fetch B` comes along and also requests it, it relies on `fetch A`'s request for `block 1` succeeding by adding it to `blockRequests`.  But if `fetch A` aborts its request, `fetch B` will never have its request completed.  This is the reason for the chunk of clean-up code here: https://github.com/geotiffjs/geotiff.js/blob/5cab63a5cf0aa963a5efa73d39804f173d7ac6d9/src/source.js#L293-L321
The other thing that is needed to really make this work is: https://github.com/geotiffjs/geotiff.js/blob/5cab63a5cf0aa963a5efa73d39804f173d7ac6d9/src/source.js#L230-L235
  which is really so that the aborting happens when the application expects it to.  In some sense this is the inverse of the example I just gave.  If `fetch B` aborted because it was no longer needed but `fetch A` was the one actually managing `block 1`, `fetch B` would have to wait for `fetch A` to complete its request for `block 1` which defeats the purpose of `fetch B`'s abort - that is, `fetch B` will be stuck at `await Promise.all(missingRequests);` because its request for `block 1` never actually aborted even though `fetch B`'s abort was signalled.

I think this covers the reasons for the changes from my PR....it's a lot, I know.  You can undo the changes in this PR for signal and then I can make a separate PR with detailed comments once this is merged if that is easier.  Let me know.

Reply:
P.S I had to run `npm uninstall geotiff` and then `npm install https://github.com/geotiffjs/geotiff.js#07ed1013` to get it to work with our repo - changing the package.json entry to `geotiffjs/geotiff.js#multi-ranges` built an old version of this branch for some reason.

Reply:
@ilan-gold Thanks for the response!

Unfortunately I did not get the viv client to work, I think I'm missing some dependency:
```bash
$ npm start

> @hms-dbmi/viv@0.9.2 start
> concurrently "npm run build:w" "cd avivator && npm start"

sh: concurrently: command not found
npm ERR! code 127
npm ERR! path /Users/fabian/dev/viv
npm ERR! command failed
npm ERR! command sh -c concurrently "npm run build:w" "cd avivator && npm start"

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/fabian/.npm/_logs/2021-02-22T17_30_31_194Z-debug.log
```

I think I understand the issue you described, and the way I'm trying to ship around it is to first request the initial set of blocks or use the ones already being requested. If any of the already being requested failed, and a signal was the cause and the signal is not the current one then I will restart just those blocks and wait for them.

This i done in dd3b36f.

Do you think this is a viable strategy? I'm a bit hesitant to juggle around signals too much, apart from checking equality and status, but not combining signals and the like. Please let me know your thoughts on that.

Reply:
I did get a bit further with viv and now get this error: 
```
[1] ERROR in ./src/utils.js 2:0-54
[1] Module not found: Error: Can't resolve 'geotiff' in '/Users/fabian/dev/viv/avivator/src'
[1]  @ ./src/index.js 8:0-41 43:19-33
[1]
[1] ERROR in ../dist/index.js 11:0-44
[1] Module not found: Error: Can't resolve 'geotiff' in '/Users/fabian/dev/viv/dist'
[1]  @ ./src/Avivator.js 8:0-87 98:19-34 337:82-98 353:40-62
[1]  @ ./src/index.js 7:0-34 47:40-48 56:38-46
[1]
[1] webpack 5.3.2 compiled with 2 errors in 10071 ms
[1] ℹ ｢wdm｣: Failed to compile.
```

Reply:
@constantinius I had a bit of trouble installing the branch too, which I had never had trouble with in the past.  Only the cli `npm install https://github.com/geotiffjs/geotiff.js#07ed1013` seemed to work for me in the top-level of the repo.   Another option would just be to copy and paste the contents of `src` from this branch into the `node_modules/geotiff/src` since Viv uses `src` directly.  I'll test out your recent changes now.

Reply:
(Let me know if you need more help, I'm happy to assist - if it seems like a bug is causing this, feel free to open an issue)

Reply:
I do think debugging in viv is helpful since it really stretches this functionality to its limit.